commit.sh runs clang-tidy before module BMIs are built → 'module std not found' on import-std TUs (blocks #326)

## Problem

When a translation unit uses `import std;` (the C++26 named std module), `clang-tidy` fails to find the std module:

```
error: module 'std' not found [clang-diagnostic-error]
```

or, when it does pick up a prebuilt PCM via compile_commands.json:

```
error: module 'std' in precompiled file '.../module-cache/.../std-*.pcm'
(imported by precompiled file 'CMakeFiles/storm.dir/storm.pcm') is not defined
in any loaded module map file; maybe you need to load '.../module.modulemap'?
```

`clang-tidy` replays `compile_commands.json`, but it does **not** reconstruct CMake's
`CXX_MODULE_STD` plumbing — i.e. the synthesized `__cmake_cxx_std_26` target that builds
`std.pcm` from libc++'s `std.cppm`. Without that, tidy has no module map / BMI for `std`, so
every `import std;` TU errors out. This is a clang-tidy / build-integration gap, **not** a
code defect.

## Why this matters

The pre-commit hook's `clang-tidy --diff --fix` step (and the CI tidy job) will fail on
**every** import-std TU. This blocks the `import std;` migration (#326): the tidy gate can't
pass once any target consumes the std module.

## Evidence

Surfaced during the #326 feasibility spike (`spike/import-std-bigbang-src`, commit e21cfb8).
After converting the whole tree to `import std;`:
- format ✓, tests ✓ (1924/1924), coverage ✓ (100%)
- **clang-tidy ✗** — `module 'std' not found` on src/, tests/, benchmarks/ TUs
- (release build ✗ was a separate stale-cache issue, see #326 docs)

Full write-up: `docs/superpowers/plans/2026-05-28-import-std-bigbang-src-spike.md` (Finding C).

## Possible approaches (to investigate)

1. **Point tidy at the prebuilt `std.pcm`** — pass the std module's BMI / module-map to
   clang-tidy explicitly (extra args in `scripts/run_clang_tidy.sh`), so it loads the same
   module CMake built.
2. **Pre-build the std module before tidy runs** and inject `-fmodule-file=std=<path>` into
   the tidy invocation.
3. **Skip import-std TUs** from the tidy gate (last resort — loses coverage; conflicts with
   the project's no-silent-fallback policy).
4. **Wait for upstream** clang-tidy module support to mature (track, don't block on).

## Definition of done

- [ ] `scripts/run_clang_tidy.sh` (and the CI tidy job) runs clean on a TU that does `import std;`
- [ ] Pre-commit `clang-tidy --diff` passes on the #326 migration branch
- [ ] Approach documented in `docs/development/` (and `COMPILER_ISSUES.md` if it's a workaround)

## Related

- Blocks #326 (single `import std;` migration)

## Investigation result (time-boxed) — root cause corrected

**clang-tidy is NOT fundamentally unable to handle `import std;`.** It works fine on import-std TUs *as long as the module BMIs exist*. The original failure was a **build-ordering bug**, compounded by Finding D (stale release dir).

### Reproduction

With a valid release build (std.pcm + storm.pcm present), tidy runs clean:
```
$ ../clang-p2996/build/bin/clang-tidy -p build/release src/db/concept.cppm
712 warnings generated.   # no "module 'std' not found"
```

Remove the std BMI and it fails exactly like the hook did — and tidy does NOT rebuild it:
```
$ mv build/release/CMakeFiles/__cmake_cxx_std_26.dir/std.pcm /tmp/
$ clang-tidy -p build/release src/db/concept.cppm
error: module file 'CMakeFiles/__cmake_cxx_std_26.dir/std.pcm' not found [clang-diagnostic-error]
    5 | import std;
```

### Root cause

In `commit.sh` the steps run in this order:
- **Step 3 (line 214): clang-tidy** — reads `build/release/compile_commands.json`, whose
  commands reference `std.pcm` / `storm.pcm`.
- **Step 3b (line 231): release build** — the step that actually *produces* those BMIs.

So tidy runs **before** the build that creates the module files it depends on. For
pre-module code this was harmless; with `import std;` (and `import storm;`) tidy now needs the
BMIs to exist first. A fresh or stale `build/release` (Finding D) means no BMIs → `module file
not found` on every import-std TU.

### Fix direction (small, no clang-tidy magic needed)

Ensure the module BMIs are built **before** Step 3. Options:
1. Move the release build (Step 3b) ahead of clang-tidy (Step 3), OR
2. Before tidy, build just the module BMIs: `cmake --build --preset ninja-release --target storm`
   (this also builds the `__cmake_cxx_std_26` std.pcm dependency).

No `-fmodule-file` injection or upstream tooling change is required. The existing
`run_clang_tidy.sh` module-error grep-filter is also unnecessary for this case once BMIs exist.

### Scope adjustment

This reduces #330 from "make clang-tidy module-aware" to "fix commit.sh step ordering so module
BMIs are built before tidy." Much smaller. Still a real prerequisite for the #326 tidy gate.

--- a/commit.sh
+++ b/commit.sh
@@ -32,6 +32,7 @@ TOTAL_START=$SECONDS
 LOG_FORMAT=/tmp/storm_format.log
 LOG_CMAKE_FORMAT=/tmp/storm_cmake_format.log
 LOG_TIDY=/tmp/storm_tidy.log
+LOG_TIDY_BMI=/tmp/storm_tidy_bmi.log
 LOG_RELEASE=/tmp/storm_release.log
 LOG_TESTS=/tmp/storm_tests.log
 LOG_COVERAGE=/tmp/storm_coverage.log
@@ -184,6 +185,15 @@ TOTAL_STEPS=0
 [[ "$RUN_FORMAT" == true ]] && ((TOTAL_STEPS++))
 [[ "$RUN_CMAKE_FORMAT" == true ]] && ((TOTAL_STEPS++))
 [[ "$RUN_TIDY" == true ]] && ((TOTAL_STEPS++))
+# clang-tidy needs module BMIs built first when the tree consumes `import std;`
+# (issue #330). Only counts as a step when build/release actually references the
+# std module — keeps the count accurate on pre-#326 trees that skip it.
+RUN_TIDY_BMI=false
+if [[ "$RUN_TIDY" == true && -f "build/release/compile_commands.json" ]] \
+   && grep -q -- '-fmodule-file=std=' "build/release/compile_commands.json" 2>/dev/null; then
+    RUN_TIDY_BMI=true
+    ((TOTAL_STEPS++))
+fi
 [[ "$RUN_BENCH_RELEASE" == true ]] && ((TOTAL_STEPS++))
 [[ "$RUN_TESTS" == true ]] && ((TOTAL_STEPS++))
 [[ "$RUN_COVERAGE" == true ]] && ((TOTAL_STEPS++))
@@ -216,6 +226,23 @@ fi
 # lines, so pre-existing drift in unrelated files doesn't block unrelated work.
 # Set STORM_TIDY_FULL=1 to force whole-file staged scan (the pre-#262 behavior).
 if [[ "$RUN_TIDY" == true ]]; then
+    # clang-tidy parses each TU using build/release/compile_commands.json. A TU
+    # that does `import std;` (or `import storm;`, which transitively imports
+    # std) needs the std/storm module BMIs (.pcm) to exist, and clang-tidy will
+    # NOT build them itself — it fails with "module file '…std.pcm' not found".
+    # Those BMIs are produced by the release build, which runs AFTER tidy
+    # (Step 3b), so on a fresh/stale build/release the tidy step fails. Build the
+    # module BMIs first — but only when the project actually consumes the std
+    # module (post-#326). On a tree without `import std;`, TUs parse standalone
+    # and this build is pure overhead, so we skip it. See issue #330.
+    if [[ "$RUN_TIDY_BMI" == true ]]; then
+        if [[ ! -f "build/release/build.ninja" ]]; then
+            cmake --preset ninja-release > /dev/null 2>&1
+        fi
+        run_step "module BMIs (for clang-tidy)" "$LOG_TIDY_BMI" \
+            cmake --build --preset ninja-release --target storm
+    fi
+
     if [[ -n "$STORM_TIDY_FULL" ]]; then
         run_step "clang-tidy --full --fix" "$LOG_TIDY" ./scripts/run_clang_tidy.sh --full --fix || true
     else

